src/NexusMods.DataModel/Abstractions/EntityDictionary.cs

```diff
@@ -238,6 +238,35 @@ public TState Walk<TState>(Func<TState, Entity, TState> visitor, TState initial)
         });
     }
 
+    /// <summary>
+    /// Merges two dictionaries together, returning a new dictionary, using mergeFn to combine results, return
+    /// null to exclude the item from the resulting dictionary.
+    /// </summary>
+    /// <param name="other"></param>
+    /// <returns></returns>
+    public EntityDictionary<TK, TV> Merge(EntityDictionary<TK, TV> other, Func<TV?, TV?, TV?> mergeFn)
+    {
+        var result = Empty(_store);
+        foreach (var (key, value) in _coll)
+        {
+            var otherValue = other._coll.TryGetValue(key, out var found) ? _store.Get<TV>(found) : null;
+            var newValue = mergeFn(_store.Get<TV>(value), otherValue);
+            if (newValue is not null)
+                result = result.With(key, newValue);
+        }
+
+        foreach (var (key, value) in other._coll)
+        {
+            // We've already processed colliding keys, so skip them.
+            if (_coll.ContainsKey(key)) continue;
+            var newValue = mergeFn(null, _store.Get<TV>(value));
+            if (newValue is not null)
+                result = result.With(key, newValue);
+        }
+
+        return result;
+    }
```

Comment:
`result = result.With(key, newValue)` uses `ImmutableDictionary.SetItem`. Doing this inside two for-each loops and creating new immutable dictionaries on possible every iteration is not very efficient. `ImmutableDictionary.SetItems` can be used instead.